sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollectionViews.java

```diff
@@ -960,6 +995,179 @@ public void verifyDeterministic() throws NonDeterministicException {
     }
   }
 
+  /**
+   * Implementation which is able to adapt an iterable materialization to a {@code List<T>}.
+   *
+   * <p>Unlike ListViewFn2, this implementation is optimized for iteration rather than indexing.
+   *
+   * <p>For internal use only.
+   */
+  public static class ListViewFn3<T> extends ViewFn<IterableView<T>, List<T>> {
```

Comment:
I don't _love_ that these are just named 1, 2, 3...

Reply:
Yes, this is equivalent to View.asIterable() plus implementing the List methods so nothing breaks. Good question about the Window -> Iterable map; this is handled at a lower level, but I don't know all the details there (though in that case I can see that constructing the mapping would be more worthwhile). In the interest of being conservative while capturing the most important gains I'll restrict this to the global window case.